File: ToV/python/agent.py
import json

import numpy as np

from feature_extractor import FeatureExtractor
from network import MLP

VERDICTS = ['GUILTY', 'INNOCENT', 'ABSTAIN']


class Agent:
    def __init__(self, game_config, hidden_vote=(64, 32), hidden_night=(48, 24),
                 hidden_claim=(48, 24), seed=None):
        self.game_config = game_config
        self.extractor = FeatureExtractor(game_config)
        self.vote_net = MLP(self.extractor.vote_dim, hidden_vote, 3, seed)
        self.night_net = MLP(self.extractor.night_dim, hidden_night, 3, seed)
        self.claim_net = MLP(self.extractor.claim_dim, hidden_claim,
                             game_config.role_count, seed)
        self.fitness = 0.0
        self.games_played = 0

    def decide_vote(self, state, memories, voter_id, candidates):
        if not candidates:
            return None
        best = None
        best_score = float('-inf')
        for c in candidates:
            f = self.extractor.vote_feature(state, memories, voter_id, c)
            p = self.vote_net.predict(f)
            score = p[0] - p[1]
            if score > best_score:
                best_score = score
                best = c
        return best if best_score > 0 else None

    def vote_verdict(self, state, memories, voter_id, accused_id):
        f = self.extractor.vote_feature(state, memories, voter_id, accused_id)
        p = self.vote_net.predict(f)
        return VERDICTS[int(np.argmax(p))]

    def decide_night_target(self, state, memories, actor_id, valid_targets):
        if not valid_targets:
            return None
        best = None
        best_score = float('-inf')
        for t in valid_targets:
            f = self.extractor.night_feature(state, memories, actor_id, t)
            p = self.night_net.predict(f)
            score = p[0] - p[2]
            if score > best_score:
                best_score = score
                best = t
        return best

    def decide_claim(self, state, memories, player_id):
        f = self.extractor.claim_feature(state, memories, player_id)
        p = self.claim_net.predict(f)
        idx = int(np.argmax(p))
        return self.game_config.role_list[idx]

    def mutate(self, rate):
        self.vote_net.mutate(rate)
        self.night_net.mutate(rate)
        self.claim_net.mutate(rate)
        return self

    def crossover(self, other):
        child = Agent(self.game_config)
        child.vote_net = self.vote_net.crossover(other.vote_net)
        child.night_net = self.night_net.crossover(other.night_net)
        child.claim_net = self.claim_net.crossover(other.claim_net)
        return child

    def clone(self):
        c = Agent(self.game_config)
        c.vote_net = self.vote_net.clone()
        c.night_net = self.night_net.clone()
        c.claim_net = self.claim_net.clone()
        return c

    def avg_fitness(self):
        return self.fitness / self.games_played if self.games_played else 0.0

    def to_dict(self):
        cfg = self.game_config
        nets = {
            'vote_net': self.vote_net.to_dict(),
            'night_net': self.night_net.to_dict(),
            'claim_net': self.claim_net.to_dict(),
        }
        return {
            'version': 1,
            'game_config': {
                'player_count': cfg.player_count,
                'preset_id': cfg.preset_id,
                'feature_dim': cfg.base_dim,
                'candidate_dim': cfg.candidate_dim,
                'role_ids': cfg.role_list,
            },
            'network_configs': dict(nets),
            'trained_weights': dict(nets),
        }

    @staticmethod
    def from_dict(data, game_config):
        if 'trained_weights' in data:
            data = data['trained_weights']
        agent = Agent(game_config)
        agent.vote_net = MLP.from_dict(data['vote_net'])
        agent.night_net = MLP.from_dict(data['night_net'])
        agent.claim_net = MLP.from_dict(data['claim_net'])
        return agent

    def save(self, path):
        with open(path, 'w') as f:
            json.dump(self.to_dict(), f)

    @staticmethod
    def load(path, game_config):
        with open(path) as f:
            return Agent.from_dict(json.load(f), game_config)

File: js/sim/agent.js
'use strict';

const engine = require('../../js/engine.js');
const { NeuralNetwork, createInputEncoder } = require('./neural-net.js');
const FeatureExtractor = require('./feature-extractor.js');

const HIDDEN_DIMS = [32, 16];
const CLAIM_HIDDEN_DIMS = [16, 8];

function probeState(playerCount) {
  return { playerCount: playerCount, players: [], graveyard: [], dayNumber: 0, night: { number: 0 } };
}

function inputDims(playerCount) {
  const probe = probeState(playerCount);
  return {
    base: FeatureExtractor.getFeatureDimension(probe),
    pair: FeatureExtractor.getVoteDimension(probe),
    roles: FeatureExtractor.getRoleList().length
  };
}

function argmax(values) {
  let best = 0;
  for (let i = 1; i < values.length; i += 1) {
    if (values[i] > values[best]) best = i;
  }
  return best;
}

class Agent {
  constructor(playerCount, role) {
    this.playerCount = playerCount;
    this.role = role;
    this.encoder = createInputEncoder(playerCount);

    const dims = inputDims(playerCount);
    this.voteNet = new NeuralNetwork([dims.pair].concat(HIDDEN_DIMS, [3]));
    this.nightNet = new NeuralNetwork([dims.pair].concat(HIDDEN_DIMS, [3]));
    this.claimNet = new NeuralNetwork([dims.base].concat(CLAIM_HIDDEN_DIMS, [dims.roles]));

    this.fitness = 0;
    this.gamesPlayed = 0;
  }

  decideVote(state, memories, voterId, candidates) {
    if (!candidates.length) return null;

    let bestScore = -Infinity;
    let bestCandidate = null;

    candidates.forEach(function (candidateId) {
      const input = this.encoder.getVoteInput(state, memories, voterId, candidateId);
      const output = this.voteNet.predict(input);
      const score = output[0] - output[1];

      if (score > bestScore) {
        bestScore = score;
        bestCandidate = candidateId;
      }
    }.bind(this));

    return bestCandidate;
  }

  getVoteVerdict(state, memories, voterId, accusedId) {
    const input = this.encoder.getVoteInput(state, memories, voterId, accusedId);
    const output = this.voteNet.predict(input);
    return ['GUILTY', 'INNOCENT', 'ABSTAIN'][argmax(output)];
  }

  decideNightTarget(state, memories, actorId, validTargets) {
    if (!validTargets.length) return null;

    let bestScore = -Infinity;
    let bestTarget = null;

    validTargets.forEach(function (targetId) {
      const input = this.encoder.getNightInput(state, memories, actorId, targetId);
      const output = this.nightNet.predict(input);
      const score = output[0] - output[2];

      if (score > bestScore) {
        bestScore = score;
        bestTarget = targetId;
      }
    }.bind(this));

    return bestTarget;
  }

  getNightAction(state, memories, actorId, validTargets) {
    if (!validTargets || !validTargets.length) return 'no_action';
    const target = this.decideNightTarget(state, memories, actorId, validTargets);
    if (target == null) return 'no_action';
    const input = this.encoder.getNightInput(state, memories, actorId, target);
    const output = this.nightNet.predict(input);
    return ['target_aggressive', 'target_defensive', 'no_action'][argmax(output)];
  }

  decideClaim(state, memories, playerId) {
    const input = this.encoder.getClaimInput(state, memories, playerId);
    const output = this.claimNet.predict(input);
    return FeatureExtractor.getRoleList()[argmax(output)] || null;
  }

  getFitness() {
    return this.fitness;
  }

  resetFitness() {
    this.fitness = 0;
    this.gamesPlayed = 0;
  }

  mutate(rate) {
    this.voteNet.mutate(rate);
    this.nightNet.mutate(rate);
    this.claimNet.mutate(rate);
  }

  crossover(other) {
    const child = new Agent(this.playerCount, this.role);
    child.voteNet = this.voteNet.crossover(other.voteNet);
    child.nightNet = this.nightNet.crossover(other.nightNet);
    child.claimNet = this.claimNet.crossover(other.claimNet);
    return child;
  }

  serialize() {
    const dims = inputDims(this.playerCount);
    return {
      version: 1,
      game_config: {
        player_count: this.playerCount,
        feature_dim: dims.base,
        candidate_dim: FeatureExtractor.getCandidateDimension(probeState(this.playerCount)),
        role_ids: FeatureExtractor.getRoleList().slice()
      },
      network_configs: {
        vote_net: this.voteNet.serialize(),
        night_net: this.nightNet.serialize(),
        claim_net: this.claimNet.serialize()
      },
      trained_weights: {
        vote_net: this.voteNet.serialize(),
        night_net: this.nightNet.serialize(),
        claim_net: this.claimNet.serialize()
      }
    };
  }

  static deserialize(data) {
    if (data && data.trained_weights && data.game_config) {
      const agent = new Agent(data.game_config.player_count, 'random');
      agent.voteNet = NeuralNetwork.deserialize(data.trained_weights.vote_net);
      agent.nightNet = NeuralNetwork.deserialize(data.trained_weights.night_net);
      agent.claimNet = NeuralNetwork.deserialize(data.trained_weights.claim_net);
      return agent;
    }
    const agent = new Agent(data.playerCount, data.role);
    agent.voteNet = NeuralNetwork.deserialize(data.voteNet);
    agent.nightNet = NeuralNetwork.deserialize(data.nightNet);
    agent.claimNet = NeuralNetwork.deserialize(data.claimNet);
    agent.fitness = data.fitness;
    agent.gamesPlayed = data.gamesPlayed;
    return agent;
  }
}

class Population {
  constructor(size, playerCount) {
    this.size = size;
    this.playerCount = playerCount;
    this.agents = [];
    this.generation = 0;

    for (let i = 0; i < size; i++) {
      this.agents.push(new Agent(playerCount, 'random'));
    }
  }

  getBest() {
    return this.agents.reduce(function (best, agent) {
      return agent.getFitness() > best.getFitness() ? agent : best;
    });
  }

  getAverage() {
    const total = this.agents.reduce(function (sum, agent) {
      return sum + agent.getFitness();
    }, 0);
    return total / this.agents.length;
  }

  select() {
    const sorted = this.agents.slice().sort(function (a, b) {
      return b.getFitness() - a.getFitness();
    });

    const survivors = sorted.slice(0, Math.floor(this.size / 2));

    const children = [];
    while (children.length < this.size) {
      const parent1 = survivors[Math.floor(Math.random() * survivors.length)];
      const parent2 = survivors[Math.floor(Math.random() * survivors.length)];
      const child = parent1.crossover(parent2);
      child.mutate(0.1);
      children.push(child);
    }

    this.agents = children;
    this.generation++;
  }

  serialize() {
    return {
      size: this.size,
      playerCount: this.playerCount,
      generation: this.generation,
      agents: this.agents.map(function (a) { return a.serialize(); })
    };
  }

  static deserialize(data) {
    const pop = new Population(data.size, data.playerCount);
    pop.generation = data.generation;
    pop.agents = data.agents.map(function (a) { return Agent.deserialize(a); });
    return pop;
  }
}

module.exports = { Agent, Population, inputDims };
